File: Java/ExportAndImportSample/src/main/java/com/thoughtspot/ExportAndImportSample/StartHere.java
package com.thoughtspot.ExportAndImportSample;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FilenameFilter;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

import org.json.simple.JSONObject;

import com.google.gson.Gson;
import com.google.gson.internal.LinkedTreeMap;

import localhost.RESTAPISDKClient;

/**
 * Test Export and Import sample.
 */
public final class StartHere {

    //Provide the host and credentials for login
    /**
     * Base URL to be connected.
     */
    private static final String BASEURL = "https://your-testcluster.thoughtspot.com";
    /**
     * Username of the user account.
     */
    private static final String USERNAME = "testuser";
    /**
     * Password of the user.
     */
    private static final String PASSWORD = "testpassword";


    /**
     * Private Constructor.
     */
    private StartHere() {
    }

    /**
     * Export TML to a file.
     *
     * @param result tml result.
     * @param filePath file path to export.
     */
    public static void exportTMLToFile(Object result, String filePath) {
        //Read the filename from info.filename and edoc fields from the response
        Gson gson = new Gson();
        JSONObject jsonObject = gson.fromJson(result.toString(), JSONObject.class);
        ArrayList<LinkedTreeMap> list = (ArrayList) jsonObject.get("object");

        for (LinkedTreeMap map : list) {
            LinkedTreeMap info = (LinkedTreeMap) map.get("info");
            String fileName = "";
            if (info.get("filename") != null) {
                fileName = info.get("filename").toString();
            } else {
                fileName = info.get("name").toString();
            }

            String edoc = "";
            if (map.get("edoc") != null) {
                edoc = (String) map.get("edoc");
            }

            FileOutputStream fos = null;
            try {
                File file = new File(filePath + "/" + fileName);
                if (file.getParentFile() != null) {
                    file.getParentFile().mkdirs();
                }
                //Write the object to file
                fos = new FileOutputStream(file);
                System.out.println(filePath + "/" + fileName);

                byte[] bytesArray = edoc.getBytes();

                fos.write(bytesArray);
                fos.flush();
                System.out.println("File Written Successfully");
            } catch (IOException ioe) {
                ioe.printStackTrace();
            } finally {
                try {
                    if (fos != null) {
                        fos.close();
                    }
                } catch (IOException ioe) {
                    System.out.println("Error in closing the Stream");
                }
            }
        }
    }

    /**
     * Get the TMLs from the file.
     *
     * @param filePath filePath.
     * @return List of TMLs which are read from file.
     * @throws IOException .
     */
    public static List<String> importTMLFromFile(String filePath) throws IOException {
        File dir = new File(filePath);
        File[] files = dir.listFiles(new FilenameFilter() {
            @Override
            public boolean accept(File dir, String name) {
                return name.endsWith(".tml");
            }
        });

        List<String> tmlList = new ArrayList<>();

        for (File f : files) {
            String content = new String(Files.readAllBytes(f.toPath()));
            tmlList.add(content);
        }
        return tmlList;
    }

    /**
     * Export and Import execution starts here.
     *
     * @param args args.
     */
    public static void main(String[] args) {
        try {
            ControllerBase controller = new ControllerBase();
            RESTAPISDKClient client = controller.getClient(BASEURL, USERNAME, PASSWORD);

            MetadataMethods metadataController = new MetadataMethods();

            String userFilter = "tsadmin";

            //Search Liveboards created by a particular user and get the list of ids
            List<String> ids = metadataController.searchObject(client, userFilter);

            //Export TML for the list of object ids received from above call
            Object result = metadataController.exportObject(client, ids);

            //Write the TMLs into file on disk
            String filePath = "/tmp/test";
            exportTMLToFile(result, filePath);

            //Read the TMLs from a folder
            List<String> tmls = importTMLFromFile(filePath);

            //Import TML received from above call
            String resp = metadataController.importObject(client, tmls);

            System.out.println("Response: " + resp);

        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

File: Python/ExportAndImportSample/ExportAndImport/StartHere.py
#! /usr/bin/env python3
# pylint: disable= no-name-in-module, W1514,  R1732, R1702,  W1514

import glob
import json
import os

from ExportAndImportSample.ExportAndImport.ControllerBase \
    import ControllerBase
from ExportAndImportSample.ExportAndImport.MetadataMethods \
    import MetadataMethods

class StartHere:
    def __init__(self):
        # Provide the host and credentials for login
        self.v_BaseURL = "https://your-testcluster.thoughtspot.com"
        self.v_UserName = "testuser"
        self.v_Password = "testpassword"

    def TMLToFile(self, result, filePath):
        filename = ''

        # Read the filename from info.filename and edoc fields from the response
        data_dict = json.loads(result)
        jsonArray = data_dict['object']
        for data in jsonArray:
            searchInfo = []
            # dataload = json.loads(data)
            searchInfo.append(data['info'])

            if searchInfo != 0:
                for test in searchInfo:
                    try:
                        filename = ''

                        filename = test['filename']
                        filename = os.path.join(filePath, filename)
                    except:
                        print('')

                    if len(filename) != 0:
                        f = open(filename, "w")
                        try:
                            edoc = []
                            edoc.append(data['edoc'])
                        except:
                            print('')
                        if edoc != 0:
                            for element in edoc:
                                f.write(element)
                            f.close()

    def ImportAndExport(self):
        client = ControllerBase.getClient(
            self, self.v_BaseURL, self.v_UserName, self.v_Password)
        userFilter = 'tsadmin'

        # Search Liveboards created by a particular user and get the list of ids
        searchIds = MetadataMethods.SearchObject(self, client, userFilter)

        # Export TML for the list of object ids received from above call
        result = MetadataMethods.ExportObject(self, client, searchIds)

        # Write the TMLs into file on disk(Existing directory)
        filePath = 'tmp/test'
        self.TMLToFile(result, filePath)

        # Read the TMLs from a folder
        tmls = self.TMLFromFile(filePath)

        # Import TML received from above call
        importResult = MetadataMethods.ImportObject(self, client, tmls)
        print('import response::' + importResult)

    def TMLFromFile(self, filePath):
        tmls = []
        os.chdir(filePath)
        for file in glob.glob("*.tml"):
            with open(file) as f:
                content = ''
                content = f.read()
            tmls.append(content)
        return tmls


if __name__ == '__main__':
    s = StartHere()
    s.ImportAndExport()
